ddcommon-ffi/src/array_queue.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Error;
+use anyhow::Context;
+use std::ffi::c_void;
+
+#[repr(C)]
+// A simple wrapper around crossbeam_queue::ArrayQueue<*mut c_void>, which is a lock free
+// bounded multi-producer and multi-consumer (MPMC) queue.
+pub struct ArrayQueue {
+    // The actual type here should be *mut crossbeam_queue::ArrayQueue<*mut c_void>.
+    // However, cbindgen does not use the module name crossbeam_queue to generate the C header.
+    // So we use *mut c_void here and cast it to the correct type in the FFI functions.
+    inner: *mut c_void,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+}
+
+impl ArrayQueue {
+    pub fn new(
+        capacity: usize,
+        item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+    ) -> Self {
+        let internal_queue: crossbeam_queue::ArrayQueue<*mut c_void> =
+            crossbeam_queue::ArrayQueue::new(capacity);
+        let inner = Box::into_raw(Box::new(internal_queue)) as *mut c_void;
+        Self {
+            inner,
+            item_delete_fn,
+        }
+    }
+
+    pub fn take(&mut self) -> Option<Box<crossbeam_queue::ArrayQueue<*mut c_void>>> {
+        let raw = std::mem::replace(&mut self.inner, std::ptr::null_mut());
+
+        if raw.is_null() {
+            None
+        } else {
+            // # Safety: the raw pointer is not null and points to a valid memory location.
+            let queue =
+                unsafe { Box::from_raw(raw as *mut crossbeam_queue::ArrayQueue<*mut c_void>) };
+            while let Some(item) = queue.pop() {
+                // # Safety: the item is a valid memory location that can be deallocated by the
+                // item_delete_fn.
+                unsafe {
+                    (self.item_delete_fn)(item);
+                }
+            }
+            Some(queue)
+        }
+    }
+}
+
+impl Drop for ArrayQueue {
+    fn drop(&mut self) {
+        drop(self.take())
+    }
+}
+
+#[allow(unused)]
+#[repr(C)]
+pub enum ArrayQueueNewResult {
+    Ok(ArrayQueue),
+    Err(Error),
+}
+
+/// Creates a new ArrayQueue with the given capacity and item_delete_fn.
+/// The item_delete_fn is called when an item is dropped from the queue.
+#[no_mangle]
+#[must_use]
+pub extern "C" fn ddog_array_queue_new(
+    capacity: usize,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+) -> ArrayQueueNewResult {
+    let ffi_queue = ArrayQueue::new(capacity, item_delete_fn);
+    ArrayQueueNewResult::Ok(ffi_queue)
+}
+
+/// Converts a *mut ArrayQueue to a &mut crossbeam_queue::ArrayQueue<*mut c_void>.
+/// # Safety
+/// The pointer is null or points to a valid memory location allocated by array_queue_new.
+unsafe fn ddog_array_queue_ptr_to_inner<'a>(
+    queue_ptr: *mut ArrayQueue,
+) -> anyhow::Result<&'a mut crossbeam_queue::ArrayQueue<*mut c_void>> {
```

Comment:
If this function is not exported (it cannot really since it returns non C types) it probably shouldn't follow the `ddog_array_queue` prefix, because it's not very idiomatic from the rust side.

Maybe make it an associated function on the type?
```rust
impl ArrayQueue {
  fn into_inner(queue_ptr: *mut ArrayQueue) -> anyhow::Result<&'a mut crossbeam_queue::ArrayQueue<*mut c_void>> {
}

// then call like
let queue = ArrayQueue::into_inner(queue_ptr);
```
}

Reply:
Also, this function casts the ArrayQueue to an `&mut`. So, it's undefined behavior to call it concurrently. Most functions on the crossbeam::ArrayQueue only require &self anyway, so IMO you should probably have to cast functions, one creating a ref for push/pop and another one getting a ref_mut for the drop.

Reply:
Done. see `as_inner_ref()` and `take()`

Reply:
Also, changed the functions to take `&ArrayQueue` instead of `&mut`